python/tests/test_topology.py

```diff
@@ -917,89 +917,96 @@ def ts_kc_distance_incremental(ts1, ts2, lambda_=0):
     t1_vecs = KCVectors(ts1.num_samples)
     t2_vecs = KCVectors(ts2.num_samples)
 
+    t1_depths = np.zeros(ts1.num_nodes)
+    t2_depths = np.zeros(ts2.num_nodes)
+
     edge_diffs_iter_1 = ts1.edge_diffs()
     tree_iter_1 = ts1.trees(sample_lists=True)
     t1, t1_diffs = next(tree_iter_1), next(edge_diffs_iter_1)
-    update_kc_incremental(t1, t1_vecs, t1_diffs, sample_maps[0])
+    update_kc_incremental(t1, t1_vecs, t1_diffs, sample_maps[0], t1_depths)
     for t2, t2_diffs in zip(ts2.trees(sample_lists=True), ts2.edge_diffs()):
-        update_kc_incremental(t2, t2_vecs, t2_diffs, sample_maps[1])
+        update_kc_incremental(t2, t2_vecs, t2_diffs, sample_maps[1], t2_depths)
         while t1_diffs[0][1] < t2_diffs[0][1]:
             span = t1_diffs[0][1] - left
             total += norm_kc_vectors(t1_vecs, t2_vecs, lambda_) * span
 
             left = t1_diffs[0][1]
             t1, t1_diffs = next(tree_iter_1), next(edge_diffs_iter_1)
-            update_kc_incremental(t1, t1_vecs, t1_diffs, sample_maps[0])
+            update_kc_incremental(t1, t1_vecs, t1_diffs, sample_maps[0], t1_depths)
         span = t2_diffs[0][1] - left
         left = t2_diffs[0][1]
         total += norm_kc_vectors(t1_vecs, t2_vecs, lambda_) * span
 
     return total / ts1.sequence_length
 
 
-def update_kc_incremental(tree, kc, edge_diffs, sample_index_map):
+def update_kc_incremental(tree, kc, edge_diffs, sample_index_map, depths):
     _, edges_out, edges_in = edge_diffs
 
-    # Update any depths/times changed by removed edges that might not be updated
-    # by inserted edges.
-    for e in edges_out:
-        update_in_subtree_pairs(tree, kc, e.child, sample_index_map)
-
-    for e in edges_in:
-        if tree.is_leaf(e.child):
-            u = e.child
+    # Update state of detached subtrees.
+    for e in reversed(edges_out):
+        u = e.child
+        depths[u] = 0
+
+        # Only update detached subtrees that remain detached. Otherwise,
+        # they must be reattached by an incoming edge and will be
+        # updated below.
+        if tree.parent(u) == tskit.NULL:
```

Comment:
This is a bit of a shortcut instead of always updating the detached subtree. I'm kind of uneasy about it because it leaves the depths on the subtree in a bit of a bad state until it is ultimately reinserted, but we do take a bit of a performance hit otherwise for what feels like avoidable work.